src/infra/datasets/inmem/src/repos/dateset_entry_repository_inmem.rs

```diff
@@ -0,0 +1,164 @@
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use dill::{component, interface, scope, Singleton};
+use kamu_datasets::{
+    DatasetEntry,
+    DatasetEntryAliasSameError,
+    DatasetEntryNotFoundError,
+    DatasetEntryRepository,
+    DeleteEntryDatasetError,
+    GetDatasetEntryError,
+    SaveDatasetEntryError,
+    SaveDatasetEntryErrorDuplicate,
+    UpdateDatasetEntryAliasError,
+};
+use opendatafabric::{AccountID, DatasetID, DatasetName};
+use tokio::sync::RwLock;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+struct State {
+    rows: HashMap<DatasetID, DatasetEntry>,
+}
+
+impl State {
+    fn new() -> Self {
+        Self {
+            rows: HashMap::new(),
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct DatasetEntryRepositoryInMemory {
+    state: Arc<RwLock<State>>,
+}
+
+#[component(pub)]
+#[interface(dyn DatasetEntryRepository)]
+#[scope(Singleton)]
+impl DatasetEntryRepositoryInMemory {
+    pub fn new() -> Self {
+        Self {
+            state: Arc::new(RwLock::new(State::new())),
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl DatasetEntryRepository for DatasetEntryRepositoryInMemory {
+    async fn get_dataset_entry(
+        &self,
+        dataset_id: &DatasetID,
+    ) -> Result<DatasetEntry, GetDatasetEntryError> {
+        let readable_state = self.state.read().await;
+
+        let maybe_dataset_entry = readable_state.rows.get(dataset_id);
+
+        let Some(dataset_entry) = maybe_dataset_entry else {
+            return Err(DatasetEntryNotFoundError::ByDatasetId(dataset_id.clone()).into());
+        };
+
+        Ok(dataset_entry.clone())
+    }
+
+    async fn get_dataset_entries_by_owner_id(
+        &self,
+        owner_id: &AccountID,
+    ) -> Result<Vec<DatasetEntry>, GetDatasetEntryError> {
+        // TODO: PERF: Slow implementation -- to reconsider if it starts to cause us
+        //             trouble
+        let readable_state = self.state.read().await;
+
+        let dataset_entries = readable_state
+            .rows
+            .values()
+            .fold(vec![], |mut acc, dataset| {
+                if dataset.owner_id == *owner_id {
+                    acc.push(dataset.clone());
+                }
+
+                acc
+            });
+
+        if dataset_entries.is_empty() {
+            return Err(DatasetEntryNotFoundError::ByOwnerId(owner_id.clone()).into());
```

Comment:
This is unnecessarily strict - just return an empty Vec and let user decide if this is a normal situation or an error.

Reply:
https://github.com/kamu-data/kamu-cli/pull/750#discussion_r1711417572

In short: Done